NotificationService/NotificationService.swift

```diff
@@ -0,0 +1,151 @@
+                let privateKey = try obtainRawPrivateKeyFromKeyChain()
+                
+                if bestAttemptContent.title != "" {
+                    let plainText = decrypt(cipherText: bestAttemptContent.title, privateKey: privateKey)
+    
+                    bestAttemptContent.title = plainText ?? "Decryption Error"
+                }
+                
+                if bestAttemptContent.subtitle != "" {
+                    let plainText = decrypt(cipherText: bestAttemptContent.subtitle, privateKey: privateKey)
+    
+                    bestAttemptContent.subtitle = plainText ?? "Decryption Error"
+                }
+                
+                if bestAttemptContent.body != "" {
+                    let plainText = decrypt(cipherText: bestAttemptContent.body, privateKey: privateKey)
+    
+                    bestAttemptContent.body = plainText ?? "Decryption Error"
+                }
+                
+                contentHandler(bestAttemptContent)
+            } catch {
+                shouldNotHappen("Something went wrong decrypting the push notification content. Error: \(error)")
+            }
+        }
+    }
+    
+    override func serviceExtensionTimeWillExpire() {
+        // Called just before the extension will be terminated by the system.
+        // Use this as an opportunity to deliver your "best attempt" at modified content, otherwise the original push payload will be used.
+        if let contentHandler = contentHandler, let bestAttemptContent =  bestAttemptContent {
+            contentHandler(bestAttemptContent)
+        }
+    }
+    
+    /**
+     Decrypts a given `cipherText` using the RSA `privateKey`
+     
+     - Parameters:
+        - cipherText: Encrypted text that should be decrypted
+        - privateKey: RSA PrivateKey from the keychain
+     
+     */
+    private func decrypt(cipherText: String, privateKey: SecKey) -> String? {
+        let cipherData = Data(base64Encoded: cipherText)! as CFData
+        
+        var error: Unmanaged<CFError>?
+        let plaintext = SecKeyCreateDecryptedData(privateKey, .rsaEncryptionOAEPSHA256, cipherData, &error)
+        
+        guard plaintext != nil else {
+            return nil
+        }
+        
+        let data: NSData = plaintext!
+        
+        return String(data: data as Data, encoding: .utf8)
+    }
+    
+    private func checkIfPrivateKeyAlreadyExists() -> Bool {
+        do {
+            let _ = try obtainPrivateKeyFromKeyChain()
+        } catch {
+            return false
+        }
+        
+        return true
+    }
+    
+    private func obtainRawPrivateKeyFromKeyChain() throws -> SecKey {
+        var item: CFTypeRef?
+        let status = SecItemCopyMatching(privateKeyKeychainQuery as CFDictionary, &item)
+        
+        guard status == errSecSuccess && item != nil else {
+            throw RSAKeyPairError.failedObtainingPrivateKeyFromKeyChain
+        }
+        
+        return item as! SecKey
```

Comment:
isn't there a nicer (safe) way to unwrap?

Reply:
This is old keychain shit... this all isn't nice 😂

Reply:
There was a more secure way :)